drv/lpc55-sprot-server/src/handler.rs

```diff
@@ -50,213 +54,125 @@ impl Handler {
-                // The first two bytes of a SinkReq payload are the U16
-                // mod 2^16 sequence number.
-                if rx_payload.len() >= 2 {
-                    let tx_payload = tx_buf.payload_mut();
-                    tx_payload[..2].copy_from_slice(&rx_payload[..2]);
-                    tx_buf.from_existing(MsgType::SinkRsp, 2)
-                } else {
-                    Ok(tx_buf.no_payload(MsgType::SinkRsp))
-                }
-            }
-            MsgType::DumpReq => {
-                let addr =
-                    u32::from_le_bytes(rx_payload[0..4].try_into().unwrap());
+            ReqBody::IoStats => Ok(RspBody::IoStats(stats.clone())),
+            ReqBody::Sprockets(req) => {
+                // TODO: Don't unwrap!
+                Ok(RspBody::Sprockets(
+                    self.sprocket.handle_deserialized(req).unwrap_lite(),
+                ))
+            }
+            ReqBody::Dump { addr } => {
                 ringbuf_entry!(Trace::Dump(addr));
-
                 let dumper = Dumper::from(DUMPER.get_task_id());
-
-                let rval: u32 = if let Err(e) = dumper.dump(addr) {
-                    e.into()
-                } else {
-                    0
-                };
-
-                let tx_payload = tx_buf.payload_mut();
-                tx_payload[0..4].copy_from_slice(&rval.to_le_bytes());
-                tx_buf.from_existing(MsgType::DumpRsp, 4)
+                match dumper.dump(addr) {
+                    Ok(()) => Ok(RspBody::Ok),
+                    Err(e) => Err(SprotError::Dump(e)),
+                }
             }
-            MsgType::UpdSwitchDefaultImageReq => {
-                match hubpack::deserialize::<
-                    drv_sprot_api::SwitchDefaultImageHeader,
-                >(rx_payload)
-                {
-                    Ok((header, _trailing_data)) => {
-                        let rsp: UpdateRspHeader = self
-                            .update
-                            .switch_default_image(header.slot, header.duration)
-                            .map(|_| None)
-                            .map_err(|e| e.into());
-                        tx_buf.serialize(MsgType::UpdSwitchDefaultImageRsp, rsp)
+            ReqBody::Update(UpdateReq::GetBlockSize) => {
+                let size = self.update.block_size()?;
+                // Block size will always fit in a u32 on these MCUs
+                Ok(RspBody::Update(UpdateRsp::BlockSize(
+                    size.try_into().unwrap(),
+                )))
+            }
+            ReqBody::Update(UpdateReq::Prep(target)) => {
+                self.update.prep_image_update(target)?;
+                Ok(RspBody::Ok)
+            }
+            ReqBody::Update(UpdateReq::WriteBlock { block_num }) => {
+                match req.blob {
+                    Some(blob) => {
+                        let end = blob.offset + blob.size;
+                        let block = &rx_buf[blob.offset..end];
```

Comment:
(if you accept my suggestion about changing the type of `blob` (see drv-sprot-api), this code goes away...)